sklearn/ensemble/tests/test_forest.py

```diff
@@ -1795,3 +1795,22 @@ def test_mse_criterion_object_segfault_smoke_test(Forest):
     est = FOREST_REGRESSORS[Forest](n_estimators=2, n_jobs=2, criterion=mse_criterion)
 
     est.fit(X_reg, y)
+
+
+def test_random_trees_embedding_feature_names_out():
+    """Check feature names out for Random Trees Embedding."""
+    random_state = np.random.RandomState(0)
+    X = np.abs(random_state.randn(100, 4))
+    n_estimators = 10
+    hasher = RandomTreesEmbedding(
+        n_estimators=n_estimators, max_depth=1, sparse_output=False
+    ).fit(X)
+    names = hasher.get_feature_names_out()
+    assert_array_equal(
+        [
+            f"randomtreesembedding_{i+1}_{j}"
+            for i in range(n_estimators)
+            for j in [1, 2]
+        ],
+        names,
+    )
```

Comment:
I think it's best to list them all out so we know what to expect directly:

```python
    hasher = RandomTreesEmbedding(
        n_estimators=2, max_depth=2, random_state=0
    ).fit(X)
    names = hasher.get_feature_names_out()

    expected_names = [
        f"randomtreesembedding_{tree}_{leaf}"
        for tree, leaf in [
            (0, 2),
            (0, 3),
            (0, 5),
            (0, 6),
            (1, 2),
            (1, 3),
            (1, 5),
            (1, 6),
        ]
    ]
    assert_array_equal(expected_names, names)
```

Note that `max_depth=2` to have more interesting leaf indices, `n_estimators=2` to make the name output smaller, and `random_state=0` to have consistent leaf indices.